crates/rust-analyzer/src/main_loop.rs

```diff
@@ -1019,6 +1022,19 @@ impl GlobalState {
                     Some(format!("rust-analyzer/flycheck/{id}")),
                 );
             }
+            FlycheckMessage::ClearCrateDiagnostics { id, generation, package_id } => {
+                let snap = self.snapshot();
+                let Ok(Some(crate_id)) = snap.analysis.crate_with_id(&package_id) else {
+                    tracing::info!("Flycheck: could not find crate '{package_id}'");
+                    return;
+                };
+                let Ok(files) = snap.analysis.files_for(crate_id) else {
+                    return;
+                };
+                for file in files {
+                    self.diagnostics.clear_file_previous_check(id, generation, file);
+                }
```

Comment:
With our changes, I feel like it'd be better to change how we store native diagnostics, keep them in a map `HashMap<PackageId, ...>` instead of per file, then when we actually pull the diagnostics out to report them we can just iterate that and turn it into per file stuff accordingly.

Then we don't need to iterate the files of a crate which is a bit fishy (if a person removes a mod statement we lose the connection to a file and now we have stale diagnostics that wont get cleared there)

Reply:
Hmmm, removing the correlation with `FileId` really doesn't play well with the way `DiagnosticsCollection` handles accessing the diagnostics. I'm honestly having quite a bit of trouble finding a way to do that cleanly, since we need quite a bit more context than just the diagnostic to get back the exact file it applies to.

Reply:
I'll have a look

Reply:
I've implemented what I meant in (sorry for stealing your work here now): https://github.com/rust-lang/rust-analyzer/pull/18729

Please give that a review, maybe I misunderstood your problem and there are issues with that approach that I don't see

Reply:
Ha, I was definitely overthinking the initial comment ! I was very hesitant of nesting yet another level of hashmap into the diagnostics, but yeah that's just cleaner that way.